subprojects/execution/src/main/java/org/gradle/internal/execution/UnitOfWork.java

```diff
@@ -150,17 +150,17 @@ default void visitDestroyable(File destroyableRoot) {}
     }
 
     /**
-     * Handles when an input cannot be read while fingerprinting.
+     * Decorate input file fingerprinting errors when appropriate.
      */
-    default void handleUnreadableInputs(InputFileFingerprintingException ex) {
-        throw ex;
+    default RuntimeException decorateInputFileFingerprintingException(InputFileFingerprintingException ex) {
```

Comment:
Is there a way to get rid of the two decorate methods from `UnitOfWork` altogether?

Reply:
They are implemented for tasks only in order to provide specialized messages for `Copy` and `Sync`. We can remove them, and simply throw in `CaptureStateBeforeExecutionStep`, but then we lose the nice error messages.

Reply:
I was wondering whether we could somehow decorate the exceptions in `TaskExecution` without having an additional API in `UnitOfWork`.

Reply:
That'd be nice, but I'm not sure I understand how you mean to do it. Want to propose a PR?

Reply:
How about this: https://github.com/gradle/gradle/pull/22049